libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -105,10 +105,13 @@ void UnusedStoreEliminator::operator()(FunctionCall const& _functionCall)
 	else
 		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
 
+	if (sideEffects.canTerminate)
+		changeUndecidedTo(State::Used, Location::Storage);
 	if (!sideEffects.canContinue)
 	{
 		changeUndecidedTo(State::Unused, Location::Memory);
-		changeUndecidedTo(sideEffects.canTerminate ? State::Used : State::Unused, Location::Storage);
+		if (!sideEffects.canTerminate)
+			changeUndecidedTo(State::Unused, Location::Storage);
```

Comment:
When is this branch hit exactly? i.e., can neither continue nor terminate

Reply:
``revert(0,0)`` :-). "canTerminate" means "can terminate the transaction successfully, i.e. while keeping state".

If the function can't continue and can't terminate, it always reverts - then all state changes that are so far undecided will uncondtionally revert and can rightly be marked as "Unused" here.

Reply:
Is `canContinue` implied by `!canTerminate`? Or when can the transaction not continue if it is guaranteed to not terminate?

Reply:
```
canContinue == true && canTerminate == true && canRevert == false

function f() {
  if calldataload(0) { return(0,0) }
}
```

```
canContinue == true && canTerminate == false && canRevert == true

function f() {
  if calldataload(0) { revert(0,0) }
}
```


```
canContinue == false && canTerminate == false && canRevert == true

function f() {
  revert(0,0)
}
```

```
canContinue == false && canTerminate == true && canRevert == false

function f() {
  return(0,0)
}
```


```
canContinue == true && canTerminate == false && canRevert == false

function f() {
}
```

Does that make it clear :-)?

Reply:
``!canContinue && !canTerminate`` "mostly" implies ``canRevert``. It's defined in ``libyul/ControlFlowSideEffects.h``.
There's one exception in which all three are false - that's infinite recursion (which effectively also reverts).

Reply:
I summarized the expected behavior for (C)ontinue, (R)evert, (T)erminate, and (I)nfinite recursion (function summary between storage writes to same slot) from the discussion so far, as follows: (0 = false, 1 = true; `?` means I'm not sure). Is this a reasonable summary?

| C  | R | T | I | Outcome
| ------------- | ------------- | --- | ---- | --- |
| 0  | 0  | 0 | 0 | Impossible |
| 0  | 0  | 0 | 1 | Both stores removed |
| 0  | 0  | 1 | 0 | Second store removed, first retained |
| 0  | 0  | 1 | 1 | Second store removed, first retained |
| 0  | 1  | 0 | 0 | Both stores removed|
| 0  | 1 | 0 | 1 | Both stores removed|
| 0  | 1  | 1 | 0 | Second store removed |
| 0  | 1  | 1 | 1 | Second store removed |
| 1  | 0  | 0 | 0 | First store removed |
| 1  | 0  | 0 | 1 | First store removed |
| 1 | 0  | 1 | 0 | Neither store removed |
| 1  | 0  | 1 | 1 | Neither store removed |
| 1  | 1  | 0 | 0 | First store removed |
| 1  | 1  | 0 | 1 | First store removed |
| 1  | 1  | 1 | 0 | Neither store removed |
| 1  | 1  | 1 | 1 | Neither store removed |